Lib/zipfile/__init__.py

```diff
@@ -1970,6 +2033,37 @@ def mkdir(self, zinfo_or_directory_name, mode=511):
             self.fp.write(zinfo.FileHeader(False))
             self.start_dir = self.fp.tell()
 
+    def _raise_if_archive_not_in_writing_mode(self, caller_name):
+        if self.mode not in ('w', 'x', 'a'):
+            raise ValueError(
+                f"{caller_name}() requires mode 'w', 'x', or 'a', but "
+                f"mode is '{self.mode}'")
+
+    def _copy_file(self, source_zipfile, source_zinfo):
```

Comment:
This is very strange. In fact, there is not much difference between these methods. What I want to say is that maybe they can be integrated (suggestion, you don't have to take it).

Reply:
While a user could call `copy_file` on multiple files, I added `copy_files` so that a source zipfile only has to be opened to be read from once, instead of opening it 10K times to read 10K files.

However, whenever a user wants to copy a single file, they could just toss that filename into a list and call `copy_files`, so `copy_file` is only around to be a bit more user-friendly. I don't know if the standard for the CPython STL is to create separate methods for operations on single and iterable parameters, or if there's a preference to only have an iterable parameter method.

I think `copy_all_files` is probably worth keeping, because without `copy_all_files`, a user would have to:
1. open the source zipfile
2. read the zinfo list
3. get the filename of each entry in the zinfo list
4. pass the list of filenames to `copy_files`

This also creates a (very slight) overhead of opening the zipfile twice.

Copying all of the contents of a zipfile might be a common use case, so I think the convenience is probably worth it. When I created a caching system with this change, it was the only method I used.

Reply:
##### Alternative API option 1:

`ZipFile.copy_from(source_zipfile, files=None)`

If `files=None`, then copy all files from the `source_zipfile`
If `files` is iterable, then copy every file in `files` 

##### Alternative API option 2:

`ZipFile.copy_from(source_zipfile, file_or_files=None)`

If `file_or_files=None`, then copy all files from the `source_zipfile`
If `file_or_files` is iterable, then copy every file in `files` 
If `file_or_files` is a string, then copy that file